procrustes/woodgate.py

```diff
@@ -0,0 +1,260 @@
+        @ E[:s, :].flatten()
+    )
+
+    if s == n:
+        D = flattened_D1.reshape(s, n)
+    else:
+        D = np.concatenate((flattened_D1, D2), axis=0)
+    return D
+
+
+def woodgate(a: np.ndarray, b: np.ndarray) -> np.ndarray:
+    r"""
+    Woodgate's algorithm for positive semidefinite Procrustes.
+
+    Parameters
+    ----------
+    a : np.ndarray
+        The matrix to be transformed.
+        This is relabelled to G as in the paper.
+    b : np.ndarray
+        The target matrix.
+        This is relabellled to F as in the paper.
+
+    Returns
+    -------
+    ProcrustesResult
+        The result of the Procrustes transformation.
+
+    Notes
+    -----
+    Given :math:`F, G \in R^{\ n\times m}`, the woodgate algorithm finds
+    :math:`P \in S^{\ n\times n}_{\geq}` such that the following is true:
+
+    .. math::
+        \text{PSDP: } min_{P} \|F - PG\|
+
+    Woodgate's algorithm takes a non-convex approach to the above problem.
+    It finds solution to the following which serves as a subroutine to our
+    original problem.
+
+    .. math::
+        \text{PSDP*: } min_{E \in R^{\ n\times n}} \|F - E'EG\|
+
+    Now, since all local minimizers of PSDP* are also global minimizers, we
+    have :math:`\hat{P} = \hat{E}'E` where :math:`\hat{E}` is any local
+    minimizer of PSDP* and :math:`\hat{P}` is the required minimizer for
+    our originak PSDP problem.
+
+    The main algorithm is as follows:
+
+    - :math:`E_0` is chosen randomly, :math:`i = 0`.
+    - Compute :math:`L(E_i)`.
+    - If :math:`L(E_i) \geq 0` then we stop and :math:`E_i` is the answer.
+    - Compute :math:`D_i`.
+    - Minimize :math:`f(E_i - w_i D_i)`.
+    - :math:`E_{i + 1} = E_i - w_i_min D_i`
+    - :math:`i = i + 1`, start from 2 again.
+
+
+    References
+    ----------
+    .. [1] Woodgate, K. G. (1996). "A new algorithm for positive semidefinite
+        procrustes". Journal of the American Statistical Association, 93(453),
+        584-588.
+    """
+
+    # We define the matrices F, G and Q as in the paper.
+    F = a
+    G = b
+    Q = F @ G.T + G @ F.T
+
+    # We define the functions f and L as in the paper.
+    L = lambda arr: (arr.T @ arr @ G @ G.T) + (G @ G.T @ arr.T @ arr) - Q
+    f = lambda arr: (1 / 2) * (
+        np.trace(F.T @ F)
+        + np.trace(arr.T @ arr @ arr.T @ arr @ G @ G.T)
+        - np.trace(arr.T @ arr @ Q)
+    )
+
+    # Main part of the algorithm.
+    i = 0
+    n = F.shape[0]
+    E = np.random.rand(n, n)
+
+    while True:
+        LE = L(E)
+        if is_pos_semi_def(LE):
+            break
+
+        LE_pos = make_positive(LE)
+        D = find_gradient(E=E, LE=LE_pos, G=G)
+
+        func = lambda w: f(E - w * D)
+        w_min = minimize(func, 1, bounds=((1e-9, None),)).x[0]
+        E = E - w_min * D
+        i += 1
+
+    print(f"Woodgate's algorithm took {i} iterations.")
+    print(f"Error = {np.linalg.norm(F - E.T @ E @ G)}.")
+    print(f"Required P = {E.T @ E}")
```

Comment:
We don't need this information in the final version and it's safe to comment them out.

Reply:
removed in 3f938e10c7d2069dbcced45d25a57b500bf60de1